### DISCUSSION TITLE
Server error - nextauth

### DISCUSSION THREAD

**Discussion by OmerAdam34** [neuefische/questions#193] [Web] [answered]
Hi, 

ich habe mein capstone-project in die quality-assurance gepostet. 
Sobald sich jedoch ein User mit seinem Github Profile einloggen möchte, wird folgende Meldung angezeigt: 
<img width="670" alt="Bildschirmfoto 2022-09-30 um 09 59 18" src="https://user-images.githubusercontent.com/107583280/193222820-d9d7c364-410d-4dac-b617-2aee1af16229.png">
Kann mir jemand sagen, weshalb der error angezeigt wird und was ggf. noch angepasst werden muss?

**Comment by actyralf** [comment, marked as answer]
This is a general problem with OAuth Providers and Vercel Preview Deployments: As the preview domains are created dynamically by Vercel, it's not possible to set up a Github or Google App beforehand with the correct domain name and callback path. A possible solution is creating a Credentials provider with hardcoded credentials that is only active in preview environments as described [here](https://github.com/nextauthjs/next-auth/issues/497). Please check this out, I will also have a deeper look into that.

**Reply by actyralf** [reply]
Vercel provides environment parameter `VERCEL_ENV` which is set to `preview` in preview environments.

You can do this in your `[...nextauth].js`:
```
import NextAuth from "next-auth";
import GithubProvider from "next-auth/providers/github";
import CredentialsProvider from "next-auth/providers/credentials";

const providers = [];
if (process.env.VERCEL_ENV === "preview") {
  providers.push(
    CredentialsProvider({
      name: "credentials",
      credentials: {
        username: { label: "Username", type: "text", placeholder: "jsmith" },
        password: { label: "Password", type: "password" },
      },
      async authorize(credentials, req) {
        if (
          credentials.username === "test" &&
          credentials.password === "test"
        ) {
          return {
            name: "Test User",
            email: "test@example.com",
          };
        } else {
          return null;
        }
      },
    })
  );
} else {
  providers.push(
    GithubProvider({
      clientId: process.env.GITHUB_ID,
      clientSecret: process.env.GITHUB_SECRET,
    })
  );
}

export const authOptions = {
  providers,
};
export default NextAuth(authOptions);
```
and this on your Login page:
```
const isPreview=process.env.VERCEL_ENV==="preview";
const LoginPage=()=>{
...
<button onClick={()=>{signIn(isPreview ? "credentials" : "github")}}
...
}
```

**Reply by OmerAdam34** [reply]
Hi Ralf, many thanks :) this helps me enormously.

can I delete "reqeset" ( async authorize(credentials, **req**)) out of my code? If I'm not wrong, this one has no use at the moment or am I wrong?

**Reply by actyralf** [reply]
Yes you can remove `req` for this scenario. If you had a real Credentials provider, the request body would contain username and password to check against...
